File: external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/KafkaConfig.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.gearpump.streaming.kafka.lib

import com.twitter.bijection.Injection
import com.typesafe.config.{ConfigFactory, Config, ConfigValue, ConfigValueFactory}
import org.apache.gearpump.streaming.kafka.lib.consumer.KafkaConsumerConfig
import org.apache.gearpump.streaming.kafka.lib.grouper.KafkaGrouperFactory
import org.apache.gearpump.streaming.kafka.lib.producer.KafkaProducerConfig
import org.apache.gearpump.streaming.transaction.api.{MessageDecoder, TimeStampFilter}
import org.apache.gearpump.util.LogUtil
import org.slf4j.Logger

import scala.util.Try

object KafkaConfig {

  val NAME = "kafkaConfig"

  val CONSUMER_TOPICS = "kafka.consumer.topics"
  val CONSUMER_EMIT_BATCH_SIZE = "kafka.consumer.emit.batch.size"
  val FETCH_THRESHOLD = "kafka.consumer.fetch.threshold"
  val FETCH_SLEEP_MS = "kafka.consumer.fetch.sleep.ms"
  val PRODUCER_TOPIC = "kafka.producer.topic"
  val PRODUCER_BATCH_SIZE = "kafka.producer.batch.size"
  val STORAGE_REPLICAS = "kafka.storage.replicas"
  val GROUPER_FACTORY_CLASS = "kafka.grouper.factory.class"
  val MESSAGE_DECODER_CLASS = "kafka.task.message.decoder.class"
  val TIMESTAMP_FILTER_CLASS = "kafka.task.timestamp.filter.class"

  def apply(gearpumpConfig: Config): KafkaConfig = new KafkaConfig(gearpumpConfig)

  def apply(resource: String): KafkaConfig = KafkaConfig(ConfigFactory.parseResources(resource))

  def apply(): KafkaConfig = KafkaConfig(ConfigFactory.empty())

  private val LOG: Logger = LogUtil.getLogger(getClass)
}

class KafkaConfig(gearpumpConfig: Config,
                  val consumerConfig: KafkaConsumerConfig = KafkaConsumerConfig(),
                  val producerConfig: KafkaProducerConfig = KafkaProducerConfig()) extends Serializable  {
  import org.apache.gearpump.streaming.kafka.lib.KafkaConfig._

  private def get[T](key: String, optValue: Option[T], defaultValue: Option[T] = None): T = {
    val value = optValue orElse defaultValue orElse (throw new RuntimeException(s"$key not found"))
    value.get
  }

  private def getString(key: String, defaultValue: Option[String] = None): String = {
    get[String](key, Try(gearpumpConfig.getString(key)).toOption, defaultValue)
  }

  private def getInt(key: String, defaultValue: Option[Int] = None): Int = {
    get[Int](key, Try(gearpumpConfig.getInt(key)).toOption, defaultValue)
  }

  private def getLong(key: String, defaultValue: Option[Long] = None): Long = {
    get[Long](key, Try(gearpumpConfig.getLong(key)).toOption, defaultValue)
  }

  private def getInstance[C](key: String, defaultValue: Option[String] = None): C = {
    Class.forName(getString(key, defaultValue)).newInstance().asInstanceOf[C]
  }

  private def configValue(anyRef: AnyRef): ConfigValue = {
    ConfigValueFactory.fromAnyRef(anyRef)
  }

  /**
   * set kafka consumer config
   * @param config kafka consumer config
   * @return new KafkaConfig based on this but with [[KafkaConsumerConfig]] set to given value
   */
  def withConsumerConfig(config: KafkaConsumerConfig): KafkaConfig = {
    new KafkaConfig(gearpumpConfig, consumerConfig = config, producerConfig)
  }

  /**
   * set kafka producer config
   * @param config kafka producer config
   * @return new KafkaConfig based on this but with [[KafkaProducerConfig]] set to given value
   */
  def withProducerConfig(config: KafkaProducerConfig): KafkaConfig = {
    new KafkaConfig(gearpumpConfig, consumerConfig, producerConfig = config)
  }

  // -------------------------------------------------------------------
  // gearpump config for kafka
  // -------------------------------------------------------------------

  /**
   * set kafka consumer topics
   * @param topics comma-separated string
   * @return new KafkaConfig based on this but with [[CONSUMER_TOPICS]] set to given value
   */
  def withConsumerTopics(topics: String): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(CONSUMER_TOPICS, configValue(topics)))
  }

  /**
   * @return a list of kafka consumer topics
   */
  def getConsumerTopics: List[String] = {
    getString(CONSUMER_TOPICS).split(",").toList
  }


  /**
   * set max number of messages to output in each `onNext` method of [[org.apache.gearpump.streaming.task.Task]]
   * @param batchSize max number of messages
   * @return new KafkaConfig based on this but with [[CONSUMER_EMIT_BATCH_SIZE]] set to given value
   */
  def withConsumerEmitBatchSize(batchSize: Int): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(CONSUMER_EMIT_BATCH_SIZE,
      configValue(Injection[Int, java.lang.Integer](batchSize))))
  }

  /**
   * @return max number of messages to output in each `onNext` method of [[org.apache.gearpump.streaming.task.Task]]
   */
  def getConsumerEmitBatchSize: Int = {
    getInt(CONSUMER_EMIT_BATCH_SIZE)
  }

  /**
   * [[org.apache.gearpump.streaming.kafka.lib.consumer.FetchThread]] will sleep for a while if no more messages or
   * the incoming queue size is above the [[FETCH_THRESHOLD]]
   * this is to set sleep interval
   * @param sleepMS sleep interval in milliseconds
   * @return new KafkaConfig based on this but with [[FETCH_SLEEP_MS]] set to given value
   */
  def withFetchSleepMS(sleepMS: Int): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(FETCH_SLEEP_MS,
      configValue(Injection[Int, java.lang.Integer](sleepMS))))
  }

  /**
   * [[org.apache.gearpump.streaming.kafka.lib.consumer.FetchThread]] will sleep for a while if no more messages or
   * the incoming queue size is above the [[FETCH_THRESHOLD]]
   * this is to get sleep interval
   * @return sleep interval in milliseconds
   */
  def getFetchSleepMS: Int = {
    getInt(FETCH_SLEEP_MS)
  }

  /**
   * [[org.apache.gearpump.streaming.kafka.lib.consumer.FetchThread]] stops fetching new messages if its incoming queue
   * size is above the threshold and starts again when the queue size is below it
   * @param threshold queue size
   * @return new KafkaConfig based on this but with [[FETCH_THRESHOLD]] set to give value
   */
  def withFetchThreshold(threshold: Int): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(FETCH_THRESHOLD,
      configValue(Injection[Int, java.lang.Integer](threshold))))
  }

  /**
   *
   * [[org.apache.gearpump.streaming.kafka.lib.consumer.FetchThread]] stops fetching new messages if its incoming queue
   * size is above the threshold and starts again when the queue size is below it
   * @return fetch threshold
   */
  def getFetchThreshold: Int = {
    getInt(FETCH_THRESHOLD)
  }

  /**
   * set kafka producer topic
   * @param topic topic string
   * @return new KafkaConfig based on this but with [[PRODUCER_TOPIC]] set to given value
   */
  def withProducerTopic(topic: String): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(PRODUCER_TOPIC, configValue(topic)))
  }

  /**
   * @return kafka producer topic in string
   */
  def getProducerTopic: String = {
    getString(PRODUCER_TOPIC)
  }


  /**
   * set kafka producer gearpumpConfig `batch.size`
   * @param size number of messages to batch
   * @return new KafkaConfig based on this but with [[PRODUCER_BATCH_SIZE]] set to given value
   */
  def withProducerBatchSize(size: Int): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(PRODUCER_BATCH_SIZE,
      configValue(Injection[Int, java.lang.Integer](size))))
  }

  /**
   * @return get kafka producer gearpumpConfig `batch.size`
   */
  def getProducerBatchSize: Int = {
    getInt(PRODUCER_BATCH_SIZE)
  }

  /**
   * set [[KafkaGrouperFactory]], whose corresponding [[org.apache.gearpump.streaming.kafka.lib.grouper.KafkaGrouper]]
   * defines how [[kafka.common.TopicAndPartition]]s are mapped to source tasks
   *
   * @param className name of the factory class
   * @return new KafkaConfig based on this but with [[GROUPER_FACTORY_CLASS]] set to given value
   */
  def withGrouperFactory(className: String): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(GROUPER_FACTORY_CLASS, configValue(className)))
  }

  /**
   * get [[KafkaGrouperFactory]] instance, whose corresponding [[org.apache.gearpump.streaming.kafka.lib.grouper.KafkaGrouper]]
   * defines how [[kafka.common.TopicAndPartition]]s are mapped to source tasks
   * @return
   */
  def getGrouperFactory: KafkaGrouperFactory = {
    getInstance[KafkaGrouperFactory](GROUPER_FACTORY_CLASS)
  }

  /**
   * set number of kafka broker replicas to persist the mapping of (offset, timestamp)
   * @param num number of replicas
   * @return new KafkaConfig based on this but with [[STORAGE_REPLICAS]] set to given value
   */
  def withStorageReplicas(num: Int): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(STORAGE_REPLICAS,
      configValue(Injection[Int, java.lang.Integer](num))))
  }

  /**
   * @return number of kafka broker replicas to persist the mapping of (offset, timestamp)
   */
  def getStorageReplicas: Int = {
    getInt(STORAGE_REPLICAS)
  }

  /**
   * set [[MessageDecoder]] to convert kafka raw bytes into gearpump [[org.apache.gearpump.Message]]
   * @param className name of decoder class
   * @return new KafkaConfig based on this but with [[MESSAGE_DECODER_CLASS]] set to given value
   */
  def withMessageDecoder(className: String): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(MESSAGE_DECODER_CLASS, configValue(className)))
  }

  /**
   * @return set [[MessageDecoder]] instance to convert kafka raw bytes into gearpump [[org.apache.gearpump.Message]]
   */
  def getMessageDecoder: MessageDecoder = {
    getInstance[MessageDecoder](MESSAGE_DECODER_CLASS)
  }

  /**
   * set [[TimeStampFilter]] to filter gearpump [[org.apache.gearpump.Message]] based on timestamp
   * @param className name of filter class
   * @return new KafkaConfig based on this but with [[TIMESTAMP_FILTER_CLASS]] set to given value
   */
  def withTimeStampFilter(className: String): KafkaConfig = {
    KafkaConfig(gearpumpConfig.withValue(TIMESTAMP_FILTER_CLASS, configValue(className)))
  }

  /**
   * @return [[TimeStampFilter]] instance to filter gearpump [[org.apache.gearpump.Message]] based on timestamp
   */
  def getTimeStampFilter: TimeStampFilter = {
    getInstance[TimeStampFilter](TIMESTAMP_FILTER_CLASS)
  }

}

File: external/kafka/src/main/java/org/apache/gearpump/streaming/kafka/util/KafkaConfig.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.gearpump.streaming.kafka.util;

import kafka.api.OffsetRequest;
import kafka.common.TopicAndPartition;
import kafka.consumer.ConsumerConfig;
import org.apache.gearpump.streaming.kafka.lib.source.DefaultKafkaMessageDecoder;
import org.apache.gearpump.streaming.kafka.lib.util.KafkaClient;
import org.apache.gearpump.streaming.kafka.lib.source.grouper.DefaultPartitionGrouper;
import org.apache.kafka.common.config.AbstractConfig;
import org.apache.kafka.common.config.ConfigDef;

import java.io.Serializable;
import java.util.Properties;


/**
 * kafka specific configs
 */
public class KafkaConfig extends AbstractConfig implements Serializable {

  private static final ConfigDef CONFIG;

  public static final String ZOOKEEPER_CONNECT_CONFIG = "zookeeper.connect";
  private static final String ZOOKEEPER_CONNECT_DOC =
      "Zookeeper connect string for Kafka topics management.";

  public static final String BOOTSTRAP_SERVERS_CONFIG = "bootstrap.servers";
  public static final String BOOTSTRAP_SERVERS_DOC = "A list of host/port pairs to use for "
      + "establishing the initial connection to the Kafka cluster. "
      + "The client will make use of all servers irrespective of which servers are specified "
      + "here for bootstrapping&mdash;this list only impacts the initial hosts used to discover "
      + "the full set of servers. This list should be in the form "
      + "<code>host1:port1,host2:port2,...</code>. Since these servers are just used for the "
      + "initial connection to discover the full cluster membership (which may change dynamically),"
      + " this list need not contain the full set of servers (you may want more than one, though, "
      + "in case a server is down).";

  public static final String CLIENT_ID_CONFIG = "client.id";
  public static final String CLIENT_ID_DOC = "An id string to pass to the server when making "
      + "requests. The purpose of this is to be able to track the source of requests beyond just "
      + "ip/port by allowing a logical application name to be included in server-side request "
      + "logging.";

  public static final String GROUP_ID_CONFIG = "group.id";
  public static final String GROUP_ID_DOC =
      "A string that uniquely identifies a set of consumers within the same consumer group";

  public static final String ENABLE_AUTO_COMMIT_CONFIG = "auto.commit.enable";
  public static final String ENABLE_AUTO_COMMIT_DOC =
      "If true the consumer's offset will be periodically committed in the background.";

  /** KafkaSource specific configs */
  public static final String CONSUMER_START_OFFSET_CONFIG = "consumer.start.offset";
  private static final String CONSUMER_START_OFFSET_DOC = "Kafka offset to start consume from. "
      + "This will be overwritten when checkpoint recover takes effect.";

  public static final String FETCH_THRESHOLD_CONFIG = "fetch.threshold";
  private static final String FETCH_THRESHOLD_DOC = "Kafka messages are fetched asynchronously "
      + "and put onto a internal queue. When the number of messages in the queue hit the threshold,"
      + "the fetch thread stops fetching, and goes to sleep. It starts fetching again when the"
      + "number falls below the threshold";

  public static final String FETCH_SLEEP_MS_CONFIG = "fetch.sleep.ms";
  private static final String FETCH_SLEEP_MS_DOC =
      "The amount of time to sleep when hitting fetch.threshold.";

  public static final String MESSAGE_DECODER_CLASS_CONFIG = "message.decoder.class";
  private static final String MESSAGE_DECODER_CLASS_DOC =
      "Message decoder class that implements the <code>MessageDecoder</code> interface.";

  public static final String PARTITION_GROUPER_CLASS_CONFIG = "partition.grouper";
  private static final String PARTITION_GROUPER_CLASS_DOC =
      "Partition grouper class that implements the <code>KafkaGrouper</code> interface.";

  public static final String REPLICATION_FACTOR_CONFIG = "replication.factor";
  public static final String REPLICATION_FACTOR_DOC =
      "The replication factor for checkpoint store topic.";

  public static final String CHECKPOINT_STORE_NAME_PREFIX_CONFIG = "checkpoint.store.name.prefix";
  public static final String CHECKPOINT_STORE_NAME_PREFIX_DOC = "Name prefix for checkpoint "
      + "store whose name will be of the form, namePrefix-sourceTopic-partitionId";

  static {
    CONFIG = new ConfigDef()
        .define(BOOTSTRAP_SERVERS_CONFIG, // required with no default value
            ConfigDef.Type.LIST,
            ConfigDef.Importance.HIGH,
            BOOTSTRAP_SERVERS_DOC)
        .define(CLIENT_ID_CONFIG,
            ConfigDef.Type.STRING,
            "",
            ConfigDef.Importance.HIGH,
            CLIENT_ID_DOC)
        .define(GROUP_ID_CONFIG,
            ConfigDef.Type.STRING,
            "",
            ConfigDef.Importance.HIGH,
            GROUP_ID_DOC)
        .define(ZOOKEEPER_CONNECT_CONFIG,
            ConfigDef.Type.STRING,
            "",
            ConfigDef.Importance.HIGH,
            ZOOKEEPER_CONNECT_DOC)
        .define(REPLICATION_FACTOR_CONFIG,
            ConfigDef.Type.INT,
            1,
            ConfigDef.Range.atLeast(1),
            ConfigDef.Importance.MEDIUM,
            REPLICATION_FACTOR_DOC)
        .define(MESSAGE_DECODER_CLASS_CONFIG,
            ConfigDef.Type.CLASS,
            DefaultKafkaMessageDecoder.class.getName(),
            ConfigDef.Importance.MEDIUM,
            MESSAGE_DECODER_CLASS_DOC)
        .define(PARTITION_GROUPER_CLASS_CONFIG,
            ConfigDef.Type.CLASS,
            DefaultPartitionGrouper.class.getName(),
            ConfigDef.Importance.MEDIUM,
            PARTITION_GROUPER_CLASS_DOC)
        .define(FETCH_THRESHOLD_CONFIG,
            ConfigDef.Type.INT,
            10000,
            ConfigDef.Range.atLeast(0),
            ConfigDef.Importance.LOW,
            FETCH_THRESHOLD_DOC)
        .define(FETCH_SLEEP_MS_CONFIG,
            ConfigDef.Type.LONG,
            100,
            ConfigDef.Range.atLeast(0),
            ConfigDef.Importance.LOW,
            FETCH_SLEEP_MS_DOC)
        .define(CONSUMER_START_OFFSET_CONFIG,
            ConfigDef.Type.LONG,
            OffsetRequest.EarliestTime(),
            ConfigDef.Range.atLeast(-2),
            ConfigDef.Importance.MEDIUM,
            CONSUMER_START_OFFSET_DOC)
        .define(ENABLE_AUTO_COMMIT_CONFIG,
            ConfigDef.Type.BOOLEAN,
            false,
            ConfigDef.Importance.MEDIUM,
            ENABLE_AUTO_COMMIT_DOC)
        .define(CHECKPOINT_STORE_NAME_PREFIX_CONFIG,
            ConfigDef.Type.STRING,
            "",
            ConfigDef.Importance.HIGH,
            CHECKPOINT_STORE_NAME_PREFIX_DOC);
  }

  public KafkaConfig(Properties props) {
    super(CONFIG, props);
  }

  public static String getCheckpointStoreNameSuffix(TopicAndPartition tp) {
    return tp.topic() + "-" + tp.partition();
  }

  public Properties getProducerConfig() {
    Properties props = new Properties();
    props.putAll(this.originals());

    // remove source properties
    removeSourceSpecificConfigs(props);

    // remove consumer properties
    removeConsumerSpecificConfigs(props);

    return props;
  }

  public String getKafkaStoreTopic(String suffix) {
    return getString(CHECKPOINT_STORE_NAME_PREFIX_CONFIG) + "-" + suffix;
  }

  public KafkaClient.KafkaClientFactory getKafkaClientFactory() {
    return KafkaClient.factory();
  }


  public ConsumerConfig getConsumerConfig() {
    Properties props = new Properties();
    props.putAll(this.originals());

    // remove source properties
    removeSourceSpecificConfigs(props);

    // remove producer properties
    removeProducerSpecificConfigs(props);

    // set consumer default property values
    if (!props.containsKey(GROUP_ID_CONFIG)) {
      props.put(GROUP_ID_CONFIG, getString(GROUP_ID_CONFIG));
    }

    return new ConsumerConfig(props);
  }

  private void removeSourceSpecificConfigs(Properties props) {
    props.remove(FETCH_SLEEP_MS_CONFIG);
    props.remove(FETCH_THRESHOLD_CONFIG);
    props.remove(PARTITION_GROUPER_CLASS_CONFIG);
    props.remove(MESSAGE_DECODER_CLASS_CONFIG);
    props.remove(REPLICATION_FACTOR_CONFIG);
    props.remove(CHECKPOINT_STORE_NAME_PREFIX_CONFIG);
  }

  private void removeConsumerSpecificConfigs(Properties props) {
    props.remove(ZOOKEEPER_CONNECT_CONFIG);
    props.remove(GROUP_ID_CONFIG);
  }

  private void removeProducerSpecificConfigs(Properties props) {
    props.remove(BOOTSTRAP_SERVERS_CONFIG);
  }


  public static class KafkaConfigFactory implements Serializable {
    public KafkaConfig getKafkaConfig(Properties props) {
      return new KafkaConfig(props);
    }
  }
}

